include/seastar/coroutine/async_generator.hh

```diff
@@ -0,0 +1,363 @@
+        handle_type::from_promise(*this).resume();
+    }
+
+    seastar::task* waiting_task() noexcept final {
+        return _waiting_task;
+    }
+
+    class element_awaiter {
+        std::remove_cvref_t<Yielded> _value;
+        constexpr bool await_ready() const noexcept {
+            return false;
+        }
+        template <typename Promise>
+        constexpr void await_suspend(std::coroutine_handle<Promise> producer) noexcept {
+            auto& current = producer.promise();
+            producer._value = std::addressof(_value);
+        }
+        constexpr void await_resume() const noexcept {}
+    };
+
+private:
+    friend class next_awaiter<Yielded>;
+};
+
+template <typename Yielded>
+class next_awaiter {
+protected:
+    async_generator_promise_base<Yielded>* _promise = nullptr;
+    std::coroutine_handle<> _producer = nullptr;
+
+    explicit next_awaiter(std::nullptr_t) noexcept {}
+    next_awaiter(async_generator_promise_base<Yielded>& promise,
+                 std::coroutine_handle<> producer) noexcept
+        : _promise{std::addressof(promise)}
+        , _producer{producer} {}
+
+public:
+    bool await_ready() const noexcept {
+        return false;
+    }
+
+    template <typename Promise>
+    std::coroutine_handle<> await_suspend(std::coroutine_handle<Promise> consumer) noexcept {
+        _promise->_consumer = consumer;
+        return _producer;
+    }
+};
+
+} // namespace internal
+
+template<typename Ref, typename Value>
+class [[nodiscard]] async_generator {
+    using value_type = std::conditional_t<std::is_void_v<Value>,
+                                          std::remove_cvref_t<Ref>,
+                                          Value>;
+    using reference_type = std::conditional_t<std::is_void_v<Value>,
+                                              Ref&&,
+                                              Ref>;
+    using yielded_type = std::conditional_t<std::is_reference_v<reference_type>,
+                                            reference_type,
+                                            const reference_type&>;
+
+public:
+    class promise_type;
+
+private:
+    using handle_type = std::coroutine_handle<promise_type>;
+    handle_type _coro = {};
+
+public:
+    class iterator;
+
+    async_generator() noexcept = default;
+    explicit async_generator(promise_type& promise) noexcept
+        : _coro(std::coroutine_handle<promise_type>::from_promise(promise))
+    {}
+    async_generator(async_generator&& other) noexcept
+        : _coro{std::exchange(other._coro, {})}
+    {}
+    async_generator(const async_generator&) = delete;
+    async_generator& operator=(const async_generator &) = delete;
+
+    ~async_generator() {
+        if (_coro) {
+            _coro.destroy();
+        }
+    }
+
+    friend void swap(async_generator& lhs, async_generator& rhs) noexcept {
+        std::swap(lhs._coro, rhs._coro);
+    }
+
+    async_generator& operator=(async_generator &&other) noexcept {
```

Comment:
nit `&&` sticks to type

Reply:
Also, better do the below inside `if (this != &other)` to protect against self-assignment;

Reply:
to check self-assignment in move ctor was a controversial topic. but i think it has been clarified by https://cplusplus.github.io/LWG/issue2468 and formalized in the standard, see https://eel.is/c++draft/utility.arg.requirements#tab:cpp17.moveassignable . 

> If t and rv do not refer to the same object, t is equivalent to the value of rv before the assignment

so, if they refer to the same object, there are no requirements on the value of the lhs of the assignment operator after the assignment expression is evaluated. hence there is no need to perform the self-assignment here.

Reply:
@tchaikov wexhave such protections against self move all over the place.
The conservative approach is to make such a move be a no-op.
This way the lvalue contains the value of itself before the move, and it remains in a valid state (though unspecifies).
Clearing self contents does not satisfy the first requiremnt of holding the state of the rvalue prior to move.

Reply:
@avikivity do we have a documented convention regarding self-move assignment semantics?
I see that `future<>` just guarantees validity but not the state after self-move is undetermined.
Maybe for control classes this is good enough, and that may be applicable in this case.
But for containers, self-move should be a no-op to prevent unpleasant surprises IMO (even though the stl isn't consistent about it).

Reply:
> @tchaikov we have such protections against self move all over the place. The conservative approach is to make such a move be a no-op. This way the lvalue contains the value of itself before the move, and it remains in a valid state (though unspecified). 

yeah, i understand the concern.

> Clearing self contents does not satisfy the first requirement of holding the state of the rvalue prior to move.

the postcondition has a requirement of "If t and rv do not refer to the same object, t is equivalent to the value of rv before the assignment", hence i am not sure we need this check in this case.

Reply:
i changed it anyway to. but personally, i think since the standard allows the undefined behavior in this case, we don't need to be conservative here.

Reply:
Let's get @avikivity's opinion on this too.
I can see cases where the generator needs to be moved around

Reply:
> I can see cases where the generator needs to be moved around

i agree. but IMHO, moving a variable into itself is not quite a typical use case.